### DISCUSSION TITLE
VS Solution Icon: use existing icon image

### DISCUSSION THREAD

**Discussion by ygoe** [ramensoftware/windhawk-mods#32] [General]
Hello,

I've just installed Windhawk and the Mod for Visual Studio Solution Icon, after having seen this system a while ago. I had used the original VS extension in VS 2017/2019 but couldn't port it to VS 2022 anymore. So I'm trying out this approach which is not subject to Visual Studio's ever-changing extensibility APIs and the troubles that come with it.

It does what it advertises, also in VS 2022. But I'd rather use the solution icons that I've added for the older extension. I think I've seen this requested somewhere before, but I wasn't able to find something like a GitHub repo for that mod. Where are these mods hosted after all? Who owns them, where are the users and where can you log in? Where are the shown ratings and how can I add one? Is Windhawk a closed universe or is it backed by some external repository or integrates authentication services? I don't get that yet.

As I see it, this mod already creates an icon from a new bitmap with text drawn to it. It should be as easy as loading an existing icon and returning that instead, right? It's been a while since I did Win32 API coding in C but I should be able to do this. I'm not sure if it could also load PNG images though but ICO files would be a start.

And next, I also need this for VS Code which is technically unable to provide this via its own extensions. Has anybody already looked into that topic? I don't understand the window subclassing code in the mod. Never did that, only created my own window classes.

**Comment by m417z** [comment]
Hello Yves,

Mark Jansen (@learn-more) is the author of the Visual Studio Solution Icon mod, as you can see [here](https://windhawk.net/mods/lm-vs-solution-icon). He's the best person to answer your mod-related questions.

Regarding the Windhawk-related questions:

> Where are these mods hosted after all?

This repository contains the official collection of Windhawk mods which appear [on the website](https://windhawk.net/mods) and in the Windhawk program. The mods here are in production state with very clear submission steps (refer to the readme). Mod development isn't done here and is up to the mod author. For example, I have my repository called [my-windhawk-mods](https://github.com/m417z/my-windhawk-mods) where I accept pull requests and can do incremental changes.

> Who owns them, where are the users and where can you log in?

The "ownership" is based on the GitHub user. Once you submit a new mod with your GitHub user, you, and only you, can submit follow-up updates. Refer to the readme in this repository.

> Where are the shown ratings and how can I add one?

You can rate the mod after you install it:

![image](https://user-images.githubusercontent.com/4129781/183258263-8cdd0c17-436a-431a-8009-de67ec0b1807.png)

> Is Windhawk a closed universe or is it backed by some external repository or integrates authentication services?

If the information above doesn't answer it, let me know and I'll try to fill the gaps.

**Comment by ygoe** [comment]
Okay thanks, I think it's clear now. I couldn't find that information or a reference to it on the website, so I didn't try to find official information elsewhere. And now I see that "learn-more" is actually a user name. I've seen that URL in the mod metadata but thought it was a placeholder or something else… 😕

**Reply by learn-more** [reply]
hey,

Author of the mod here.

> It does what it advertises, also in VS 2022. But I'd rather use the solution icons that I've added for the older extension.

This has been on my todo-list for a while, but I have been pushing it back in favor of other things.
What filenames / filetypes did you use with the original extension?

**Reply by ygoe** [reply]
Thanks for reaching back here, I couldn't create an issue in your repo because it's not enabled. (It would probably be confusing if you always have to fork the entire mods repo just to add one mod.)

I have .editoricon.png in my repositories. I think .ico didn't work well with the Visual Studio extension, but I could also use that if that's what this mod could support.

Do you know if that could also be extended to VS Code?

**Reply by learn-more** [reply]
> Do you know if that could also be extended to VS Code?

That should be possible.
The biggest hurdle for that is finding how to query VS Code for the 'active project'.

**Reply by ygoe** [reply]
Hm, I couldn't find out how to get that info of VS Code from another process. It's only in the commandline if Code has been opened with a directory directly. And I don't know what that restart manager is which is used for VS.

**Reply by learn-more** [reply]
I have added support for the .ico in #33.
When this PR is merged, you can use the updated version to show existing icons.
Sadly, reading a png is not really trivial with winapi, so I skipped that for now.

**Reply by ygoe** [reply]
Thank you very much!

This might be helpful when implementing the PNG support:
https://stackoverflow.com/questions/4567875/how-would-i-load-a-png-image-using-win32-gdi-no-gdi-if-possible
It uses Windows Imaging Component (WIC) that is actually the core of WPF's image support and available to native applications on any supported Windows version and the ones before.

**Reply by m417z** [reply]
> The biggest hurdle for that is finding how to query VS Code for the 'active project'.

That's an interesting challenge. Aside from implementing a companion VSCode extension, here's what I could think of:
There's the `window.title` setting which allows to control the window title. Its default value is:
`"${dirty}${activeEditorShort}${separator}${rootName}${separator}${appName}"`
At the price of adding slight visual noise to the title bar, the value can be changed to something like:
`"${dirty}${activeEditorShort}${separator}${rootName}${separator}${appName} <${rootPath}>"`
Then, the opened workspace can be queried from the window title.

**Reply by learn-more** [reply]
Yeah, injecting code without writing an extension does not seem trivial sadly.
I was hoping to (ab)use node's exports in code.exe to run my own code,
but that seems far from trivial.

**Reply by m417z** [reply]
Another way is hooking file reads like [VSCode Tweaker](https://windhawk.net/mods/vscode-tweaker) does, but some of the files have their checksums verified (which VSCode Tweaker adjusts). For this specific use case, it's probably easy to find a file which is being included and isn't signed.
